### ISSUE TITLE
Softmax: Large Tensor Kernel Doesn't Support In-Place Operation

### ISSUE THREAD

**Issue by edwinleeTT** [tenstorrent/tt-metal#45200]
A TT_FATAL was added in https://github.com/tenstorrent/tt-metal/pull/23130/changes#diff-bf9242eb1d0a3d9b8603503f94599e212c78cb2587e8812244e3eae1631bfacfR148 that prevents in place softmax from being run with the large tensor kernels, with the error message
`Tensor is too large to run softmax inplace, please use standard softmax`

This assertion is now found here https://github.com/tenstorrent/tt-metal/blob/5e9f894fc64029da97ef0223c566f9ad91ecfac0/ttnn/cpp/ttnn/operations/normalization/softmax/device/softmax_program_factory_attention_optimized.cpp#L126

The large tensor kernel should in theory be able to function if the output memory location is the same as the input memory location. However, removing this assertion and attempting to exercise the following code results in a hang on the first iteration on an n150.

```
@pytest.mark.parametrize("target_sequence_size", [8192])
@pytest.mark.parametrize("sequence_size", [384])
@pytest.mark.parametrize("iterations", [50])
def test_transformer_attention_softmax_inplace_large_kernel_stress(
    target_sequence_size, sequence_size, iterations, device
):
    torch.manual_seed(0)

    input_shape = (1, 1, sequence_size, target_sequence_size)

    torch_input_tensor = torch_random(input_shape, -100.0, 100.0, dtype=torch.bfloat16)
    torch_attention_mask = torch_random(
        (1, 1, sequence_size, target_sequence_size), 0, 1.0, dtype=torch.bfloat16
    )

    golden_function = ttnn.get_golden_function(ttnn.transformer.attention_softmax_)
    torch_output_tensor = golden_function(torch_input_tensor, head_size=None, attention_mask=torch_attention_mask)

    # Mask is read-only across iterations, so upload once.
    attention_mask = ttnn.from_torch(torch_attention_mask, device=device, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT)

    first_output = None
    for i in range(iterations):
        print(f"[stress] iter {i}/{iterations} starting", flush=True)

        # Re-upload the input every iteration because attention_softmax_ mutates it in place.
        input_tensor = ttnn.from_torch(
            torch_input_tensor, device=device, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT
        )
        output_tensor = ttnn.transformer.attention_softmax_(
            input_tensor, head_size=None, attention_mask=attention_mask, causal_mask=True
        )
        output_tensor = ttnn.to_torch(ttnn.from_device(output_tensor))
        assert_with_pcc(torch_output_tensor, output_tensor, 0.99)

        # Check if results are ND
        if first_output is None:
            first_output = output_tensor
        else:
            assert torch.equal(first_output, output_tensor), (
                f"In-place large-kernel softmax is non-deterministic at iteration {i}: "
                f"output differs from iteration 0 (max abs delta "
                f"{(first_output - output_tensor).abs().max().item()})"
            )
```
